rpk: add --node-ids option to rpk cluster partitions list

daisukebe opened the pull request:
When using `rpk cluster partitions list`, it's useful if we could filter the list of partitions with broker IDs. I've added a new `--node-ids` option to the command.

```
List partitions which replicas are assigned to brokers 1 and 2.

  rpk cluster partitions list foo --node-ids 1,2
```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes

### Features

* rpk: `rpk cluster partitions list` now supports filtering with broker IDs.

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -248,3 +257,39 @@ func filterPartition(cPartitions []adminapi.ClusterPartition, partitions []int)
+// clusterPartitions with the same brokers present in the Replicas slice.
+func filterBroker(cPartitions []adminapi.ClusterPartition, brokers []int) (ret []adminapi.ClusterPartition) {
+	for _, p := range cPartitions {
+		rob := replicaOnBroker(p.Replicas, brokers)
```
[question] So the filtering works in a way that we will only show partitions that have replicas in **all** the broker list that I pass, i.e:

- I have a partition `A` with a replica in broker 1
- I have a partition `B` with replicas in brokers 1 and 2

If I filter `--broker 1,2` will I only see partition `B` ?

daisukebe replied:
Yes, that's my intention. It's AND condition, not OR. Is it unclear?

r-vasquez replied:
Awesome, it's clear. I just wanted to make sure. :+1:

r-vasquez reviewed (commented):
LGTM, only a few comments, the most important one is the `--broker` flag naming :+1:

daisukebe commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -248,3 +257,39 @@ func filterPartition(cPartitions []adminapi.ClusterPartition, partitions []int)
+// exist in the Replicas slice or the brokers slice is empty.
+// Otherwise, this returns false.
+func replicaOnBroker(replicas adminapi.Replicas, brokers []int) bool {
+	if len(brokers) == 0 {
+		return true
+	}
```
You're right. My bad, thanks for catching it.

daisukebe commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -150,12 +155,16 @@ List all in json format.
 	cmd.Flags().BoolVarP(&all, "all", "a", false, "If true, list all partitions in the cluster")
 	cmd.Flags().BoolVar(&disabledOnly, "disabled-only", false, "If true, list disabled partitions only")
 	cmd.Flags().IntSliceVarP(&partitions, "partition", "p", nil, "List of comma-separated partitions IDs that you wish to filter the results with")
+	cmd.Flags().IntSliceVarP(&brokers, "brokers", "b", nil, "List of comma-separated broker IDs that you wish to filter the results with")
```
`node-ids` sounds good.

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -72,6 +73,10 @@ List all partitions in the cluster.
 List all partitions in the cluster, filtering for topic foo and bar.
   rpk cluster partitions list foo bar
 
+List partitions which replicas are assigned on specific brokers.
+
+  rpk cluster partitions list foo --brokers 1,2
```
What do you think of this to make sure that we document `--brokers` being an inclusive filter?
```suggestion
List partitions in which replicas are assigned to brokers 1 and 2.
  rpk cluster partitions list foo --brokers 1,2
```

daisukebe replied:
+1, thank you!

daisukebe commented:
* Changed the option name from `--brokers` to `--node-ids`
* Made the help text explicit the given node-ids are inclusive
* Removed the redundant check

vbotbuildovich commented:
/backport v23.3.x
